### ISSUE TITLE
NPE with watch and multi/exec operations

### ISSUE THREAD

**Issue by Helenl1987** [distributedio/titan#126]
I was trying to make multiple threads access and modify one key using check-and-set mechanism (the watch operation), and got NullPointerException when doing transaction.exec (however, actually the transaction was executed cuz the key has been incremented).

Here's a brief example of what I'm trying to do:
```
public static void syncTest(Pool<Jedis> jedisPool) {
        int retryTimes = 10;
        long threadId = Thread.currentThread().getId();
        while (retryTimes > 0) {
            try (Jedis jedis = jedisPool.getResource()) {
                jedis.watch(key);
                String value = jedis.get(key);
                if (value == null) {
                    value = "0";
                }
                int numValue = Integer.parseInt(value) + 1;

                Transaction transaction = jedis.multi();
                transaction.set(key, String.valueOf(numValue));
                List<Object> response = transaction.exec();
                transaction.clear();
                transaction.close();

                if (response != null && !response.isEmpty()) {
                    System.out.println(String.format("thread-%d: success, old-%s new-%d "
                                    + "response-%s", threadId, value, numValue, response.toString()));
                    return;
                }

                retryTimes--;
                System.out.println(String.format("thread-%d: transaction exec failed with id "
                        + "counter old-%s new-%d, retrying", threadId, value, numValue));
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            catch (NullPointerException e) {
                System.out.println(String.format("thread-%d", threadId));
                e.printStackTrace();
            }
        }
    }
```

And the stack trace:
```
thread-22
java.lang.NullPointerException
	at redis.clients.jedis.Transaction.exec(Transaction.java:53)
	at com.company.Main.syncTest(Main.java:116)
	at com.company.Main.lambda$test2$1(Main.java:72)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Can anyone help with this? Thanks!

**Comment by arthurkiller** [comment]
Seems it's the problem of `jedis`. Have you tried your job via `redis-cli`?

**Comment by shafreeck** [comment]
It threw this exception because the server returned an empty array. Redis returns an empty array when the watched was touched. So think it is normal behavior.

You can reproduce this following the sequence drawn by the screenshot below.

![image](https://user-images.githubusercontent.com/418483/61220670-64606180-a749-11e9-90aa-d1ed039f9a83.png)

It seems that this is a BUG of Jedis, to confirm that, would you like to test your case on the official Redis?

**Comment by Helenl1987** [comment]
Thanks for your kind reply!

I've retried the same code on the official redis, jedis would return an empty list when the watched key was touched, but without throwing a NullPointerException.

And actually in this case, I think the watched key wasn't touched, cuz the counter has been incremented successfully, which means the transaction has been executed instead of discarded.

So I don't think it's a problem of jedis. Please let me know if you'd like to try my test code, happy to provide more details, thanks!

**Comment by shafreeck** [comment]
Thanks, yes, I am going to try your test case. It seems more likely an issue of Titan now. 

* What's your version of Jedis and Titan?
* Send me the whole java code if don't mind, because I am familiar with Java. 😅

**Comment by Helenl1987** [comment]
Version: jedis 2.10.2, and latest titan (sorry but I can't find any version information for titan, I'm using docker-compose with the default config file titan.toml)

Please see the attachment for the java test code. To run it, you would have to start a redis server at localhost:6379, as well as titan at localhost:47369. The port field at line 59 can be modified to specify which server to connect with.

With the official redis the expected result is counter=10, without any exception. And with titan, the counter would exceed 10, and throwing several NPEs.

Thx!

[Main.txt](https://github.com/distributedio/titan/files/3395578/Main.txt)

**Comment by Helenl1987** [comment]
@shafreeck Hi, have you tried the test case and figured out if it was an issue of titan?

If yes, could you please let me know your plan/timeline for fixing this issue?

Thanks!

**Comment by shafreeck** [comment]
Sorry for forgetting this issue A😂, I will try and fix it this week.

**Comment by shafreeck** [comment]
@Helenl1987 It was confirmed as a bug of titan, thanks for your test code. It would be fixed in this PR: #136

**Comment by Helenl1987** [comment]
Great, thanks for your help!
